File: OrchidCore/src/main/java/com/eden/orchid/impl/compilers/parsers/YamlParser.java
package com.eden.orchid.impl.compilers.parsers;

import com.eden.orchid.api.compilers.OrchidParser;
import org.json.JSONArray;
import org.json.JSONObject;
import org.yaml.snakeyaml.Yaml;

import javax.inject.Inject;
import java.util.List;
import java.util.Map;

public final class YamlParser extends OrchidParser {

    @Inject
    public YamlParser() {
        super(100);
    }

    @Override
    public String getDelimiter() {
        return "-";
    }

    @Override
    public String[] getSourceExtensions() {
        return new String[] {"yaml", "yml"};
    }

    @Override
    public JSONObject parse(String extension, String input) {
        try {
            Object yamlData = new Yaml().load(input);

            if(yamlData instanceof Map) {
                return new JSONObject((Map<String, Object>) yamlData);
            }
            else if(yamlData instanceof List) {
                JSONObject object = new JSONObject();
                object.put(OrchidParser.arrayAsObjectKey, new JSONArray((List<Object>) yamlData));
                return object;
            }
        }
        catch (Exception e) {}

        return null;
    }

    @Override
    public String serialize(String extension, Object input) {
        return new Yaml().dump(input);
    }
}

File: OrchidCore/src/main/kotlin/com/eden/orchid/impl/compilers/parsers/YamlParser.kt
package com.eden.orchid.impl.compilers.parsers

import com.caseyjbrooks.clog.Clog
import com.eden.orchid.api.compilers.OrchidParser
import com.eden.orchid.api.options.annotations.Archetype
import com.eden.orchid.api.options.archetypes.ConfigArchetype
import com.eden.orchid.utilities.SuppressedWarnings
import com.eden.orchid.utilities.logSyntaxError
import org.yaml.snakeyaml.Yaml
import org.yaml.snakeyaml.parser.ParserException
import java.util.HashMap
import javax.inject.Inject

@Archetype(value = ConfigArchetype::class, key = "services.parsers.yml")
class YamlParser @Inject
constructor() : OrchidParser() {

    override fun getDelimiter(): String {
        return "-"
    }

    override fun getSourceExtensions(): Array<String> {
        return arrayOf("yml", "yaml")
    }

    @Suppress(SuppressedWarnings.UNCHECKED_KOTLIN)
    override fun parse(extension: String, input: String): Map<String, Any>? {
        try {
            val yamlData = Yaml().load<Any>(input)

            if (yamlData is Map<*, *>) {
                return yamlData as Map<String, Any>
            } else if (yamlData is List<*>) {
                val yamlObject = HashMap<String, Any>()
                yamlObject[OrchidParser.arrayAsObjectKey] = yamlData
                return yamlObject
            }
        } catch (e: ParserException) {
            input.logSyntaxError(extension, e.problemMark.line, e.problemMark.column, e.problem)
        } catch (e: Exception) {
            Clog.e("Error parsing YAML:\n{}", e.message)
        }

        return null
    }

    override fun serialize(extension: String, input: Any): String {
        return Yaml().dump(input)
    }
}
